Add enable/disable user api

rpocklin opened the pull request:
I have added `enable` and `disable` user API endpoints, these were missing and needed for some functionality I am currently working on.  I have written the tests without regard to your pact.io work in `tests/userTest.js` which tests both methods in succession.

madsi1m commented:
@DeepDiver1975 is it possible to get this merged? Happy to receive some feedback on this.

DeepDiver1975 commented:
> @DeepDiver1975 is it possible to get this merged? Happy to receive some feedback on this.

1. branch is outdated - please rebase
2. ci is red - please have a look

THX

rpocklin commented:
> > @DeepDiver1975 is it possible to get this merged? Happy to receive some feedback on this.
> 
> 1. branch is outdated - please rebase
> 2. ci is red - please have a look
> 
> THX

I have updated the branch, but 'm not sure my original message was passed on correctly.

The tests I added pass locally when I run OC on the same domain/port as the tests, but seems to have issues when the OC server is on a different port (as it is in your Drone CI environment).

I noticed you have several other tests which may have a similar issue and have been ignored eg:
https://github.com/owncloud/owncloud-sdk/blob/47e309a62c7f83f920be893b347f5517fb2277d2/tests/appsTest.js#L65

Perhaps there is some incomplete test config scaffolding, either in jasmine or Pact or something missing in the actual implementation.  To combine testing of enable and disabling of the User I perform a `login()` command which generates the CORS error as a result.  Otherwise it is just re-using existing code to make the API calls.

Can you please take a look at this locally and then give some feedback on it?

individual-it commented:
I will have a look and convert the tests to pact.io, without a new pact interaction they will not pass

individual-it commented:
@rpocklin Thank you for investing time in the tests! 
I've made a PR to your branch and converted the tests to pact.io https://github.com/rpocklin/owncloud-sdk/pull/1
I had to rebase on master, so if you rebase your branch on master only the last commit should be left.

A couple of comments:
- I deleted the `testUser.login()` steps - here we test the client library, I don't think we need to test if the server has really disabled/enabled the user, server-facing API tests should do that. (I know this kind of thing is done a lot in this SDK, need some refactoring)
- I would rather add some tests to check if the enable/disable functions react correctly in cases where the server does not reply `OK`. E.g. when trying to disable/enable a not existing user the functions seem not to throw an exception but just come back with `true`. I tried a test where the server replies with 
```
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>101</statuscode>
  <message/>
 </meta>
 <data/>
</ocs>
```
and expected the function to fail, but it does not see https://gist.github.com/individual-it/f180a247338c32a395bf70808a344847

P.S. I'm happy to open a new PR if the rebasing is too confusing (we are currently refactoring all the tests and so there are constant changes and conflicts)

rpocklin commented:
I will be looking at this again shortly...

rpocklin commented:
I will be picking this up again to work on now...

rpocklin commented:
@individual-it thanks for helping, what should be in my `config/config.json` for these tests to run?  Is copying the `config.sample.json` and updating the `owncloudURL` and username / password fields enough?

Even on `master` branch, running `test-provider` passes but `test-consumer` fails, do I need to specify a PACTFLOW_TOKEN?
Here is the summary result of `yarn test-consumer`
```
Test Suites: 18 failed, 1 skipped, 10 passed, 28 of 29 total
Tests:       130 failed, 51 skipped, 70 passed, 251 total
Snapshots:   0 total
Time:        26.32 s
Ran all test suites.
```

individual-it commented:
@rpocklin I see we need better docs here, sorry for that. I've created and issue to improve docs #758
for the consumer tests set `"owncloudURL": "http://127.0.0.1:1234/"` this is a legacy name, these days it only refers to the mock server - issue to improve: #757 

to run the provider tests locally you have to set `PROVIDER_BASE_URL` and point it to your owncloud installation e.g. `PROVIDER_BASE_URL=http://localhost/owncloud-core yarn test-provider` - some of the tests still will fail because the contract needs still some adjustment to match both the SDK & the oC10 / OCIS provider. We are currently in the process of doing that. The CI ignores those failures on the provider side currently

I believe CI fails on the provider side for you because you are running if from an own fork. I have to look into why that is

individual-it commented:
now CI passes but I think it picked it up from #759

rpocklin commented:
@individual-it Yeah it looks like it won't get through the pact tests via drone from other repos/forks?

Running it locally I get the following (after updating `owncloudURL`):

`yarn test-consumer`
```
Test Suites: 12 failed, 1 skipped, 16 passed, 28 of 29 total
Tests:       61 failed, 51 skipped, 139 passed, 251 total
Snapshots:   0 total
Time:        9.595 s, estimated 10 s
Ran all test suites.
```

individual-it commented on tests/userTest.js:
```diff
@@ -592,6 +592,136 @@ describe('Main: Currently testing user management,', function () {
+      .withRequest({
+        method: 'PUT',
+        path: MatchersV3.regex(
+          '.*\\/ocs\\/v(1|2)\\.php\\/cloud\\/users\\/' + config.testUser + '\\/disable',
```
we figured out we don't need `v(1|2)` in the regex
```suggestion
          '.*\\/ocs\\/v1\\.php\\/cloud\\/users\\/' + config.testUser + '\\/disable',
```

rpocklin replied:
Just pushed that versioning update in the `userTests.js` file

individual-it reviewed (commented):
just a small comment rest of the consumer tests look good
the provider tests don't pass because the given steps are not added, but don't worry about them yet, there are no docs yet for that and other tests are not done ether.
After merging this PR just open an issue to adjust the provider tests in this file and ping me

review of the code itself should maybe be done by @LukasHirt or @kulmann

individual-it commented:
@rpocklin the issue with pact seems to be the `/` in the branch-name. tests work from a fork if there are no `/` in the branch name see https://github.com/owncloud/owncloud-sdk/pull/785

rpocklin commented:
@individual-it ahh that's good to know moving forward.  I can't change the branch for this PR, do you want to close this and continue in https://github.com/owncloud/owncloud-sdk/pull/759?  Does it just need a code review now?

individual-it commented:
closing this in favour of #759
@kulmann @LukasHirt please review the technical details of #759 I will take care of the tests there

individual-it commented:
@rpocklin it took for ever but finally your changes got merged, thank you for your contributions and also the effort you put in to add the tests

rpocklin commented:
@individual-it No worries, that's fine - great that it got in and thanks for the collaboration :)
